vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -30,361 +35,315 @@
-            dx = cost_map.dim_size(0) - 1;
-        }
-        if (dy >= (ssize_t)cost_map.dim_size(1)) {
-            dy = cost_map.dim_size(1) - 1;
-        }
+    if (dx >= (ssize_t)cost_map.dim_size(0)) {
+        dx = cost_map.dim_size(0) - 1;
+    }
+    if (dy >= (ssize_t)cost_map.dim_size(1)) {
+        dy = cost_map.dim_size(1) - 1;
+    }
 
-        return cost_map_[from_seg_index][size_t(box_id)][dx][dy];
+    return cost_map_[from_seg_index][size_t(box_id)][dx][dy];
+}
+
+void CostMap::set_cost_map(int from_seg_index,
+                           const routing_cost_list& costs,
+                           e_representative_entry_method method) {
+    // sort the entries
+    const auto& device_ctx = g_vpr_ctx.device();
+    std::vector<routing_cost_list> costs_per_box(device_ctx.connection_boxes.num_connection_box_types());
+    for (const auto& entry : costs) {
+        costs_per_box[size_t(std::get<3>(entry))].push_back(entry);
     }
 
-    void set_cost_map(int from_seg_index,
-                      const t_routing_cost_map& cost_map,
-                      e_representative_entry_method method) {
-        const auto& device_ctx = g_vpr_ctx.device();
-        for (size_t box_id = 0;
-             box_id < device_ctx.connection_boxes.num_connection_box_types();
-             ++box_id) {
-            set_cost_map(from_seg_index, ConnectionBoxId(box_id), cost_map, method);
-        }
+    for (size_t box_id = 0;
+         box_id < device_ctx.connection_boxes.num_connection_box_types();
+         ++box_id) {
+        set_cost_map(from_seg_index, ConnectionBoxId(box_id), costs_per_box[box_id], method);
     }
+}
 
-    void set_cost_map(int from_seg_index, ConnectionBoxId box_id, const t_routing_cost_map& cost_map, e_representative_entry_method method) {
-        VTR_ASSERT(from_seg_index >= 0 && from_seg_index < (ssize_t)offset_.size());
+void CostMap::set_cost_map(int from_seg_index, ConnectionBoxId box_id, const routing_cost_list& costs, e_representative_entry_method method) {
+    VTR_ASSERT(from_seg_index >= 0 && from_seg_index < (ssize_t)offset_.size());
 
-        // Find coordinate offset for this segment.
-        int min_dx = 0;
-        int min_dy = 0;
-        int max_dx = 0;
-        int max_dy = 0;
-        for (const auto& entry : cost_map) {
-            if (std::get<3>(entry) != box_id) {
-                continue;
-            }
-            min_dx = std::min(std::get<1>(entry).first, min_dx);
-            min_dy = std::min(std::get<1>(entry).second, min_dy);
+    // calculate the bounding box
+    BoundingBox<ssize_t> bounds;
+    for (const auto& entry : costs) {
+        bounds |= BoundingBox<ssize_t>(std::get<1>(entry).first,
+                                       std::get<1>(entry).second);
+    }
 
-            max_dx = std::max(std::get<1>(entry).first, max_dx);
-            max_dy = std::max(std::get<1>(entry).second, max_dy);
-        }
+    if (bounds.empty()) {
```

Comment:
Maybe I'm missing something, but isn't `bounds.empty()` simply the same as `costs.empty()`?

Reply:
That was from before I bucketed the costs. I'll fix it.